File: scripts/generate-icon.swift
#!/usr/bin/env swift
//
// generate-icon.swift — Produce the GitRelay AppIcon asset from scratch.
//
// Run once from the repo root, on a Mac:
//     swift scripts/generate-icon.swift
//
// Writes 10 PNG files into gitrelay/Assets.xcassets/AppIcon.appiconset/
// at exact pixel dimensions (no Retina-doubling).
//
// The mark is a deep ink squircle plate carrying a white Y-shaped git branch of
// three hollow nodes (#92 / #121). Its geometry mirrors `GitRelayMark` in
// GitRelayCore/Design/GitRelayMark.swift, which is what the menu-bar status item
// draws from. A standalone `swift` script cannot import the app module, so the
// numbers below are duplicated; `GitRelayMarkTests` pins the same values so a
// change on either side fails the test suite instead of drifting silently.

import AppKit
import Foundation

let assetPath = "gitrelay/Assets.xcassets/AppIcon.appiconset"

let sizes: [(filename: String, pixels: Int)] = [
    ("icon_16.png",        16),
    ("icon_16@2x.png",     32),
    ("icon_32.png",        32),
    ("icon_32@2x.png",     64),
    ("icon_128.png",      128),
    ("icon_128@2x.png",   256),
    ("icon_256.png",      256),
    ("icon_256@2x.png",   512),
    ("icon_512.png",      512),
    ("icon_512@2x.png",  1024),
]

// MARK: - Mark geometry (mirrors GitRelayCore/Design/GitRelayMark.swift)

struct MarkPoint {
    var x: Double
    var y: Double
}

enum Mark {
    static let strokeWidth = 0.044
    static let nodeRadius = 0.072
    static var outerRadius: Double { nodeRadius + strokeWidth / 2 }
    static var innerRadius: Double { nodeRadius - strokeWidth / 2 }

    static let fork = MarkPoint(x: 0.5, y: 0.505)
    static let nodes = [
        MarkPoint(x: 0.5, y: 0.69),     // trunk
        MarkPoint(x: 0.293, y: 0.283),  // left branch
        MarkPoint(x: 0.707, y: 0.283),  // right branch
    ]

    static let plateExponent = 5.0
    static let plateSampleCount = 720

    static func edgePoint(from node: MarkPoint, towards target: MarkPoint) -> MarkPoint {
        let dx = target.x - node.x
        let dy = target.y - node.y
        let length = (dx * dx + dy * dy).squareRoot()
        guard length > 0 else { return node }
        return MarkPoint(
            x: node.x + dx / length * outerRadius,
            y: node.y + dy / length * outerRadius
        )
    }

    static func plateOutline() -> [MarkPoint] {
        let exponent = 2 / plateExponent
        return (0..<plateSampleCount).map { index in
            let angle = 2 * Double.pi * Double(index) / Double(plateSampleCount)
            return MarkPoint(
                x: 0.5 + 0.5 * signedPower(cos(angle), exponent),
                y: 0.5 + 0.5 * signedPower(sin(angle), exponent)
            )
        }
    }

    private static func signedPower(_ value: Double, _ exponent: Double) -> Double {
        let magnitude = pow(abs(value), exponent)
        return value < 0 ? -magnitude : magnitude
    }
}

// MARK: - Palette

let plateTop = NSColor(srgbRed: 0.204, green: 0.204, blue: 0.216, alpha: 1)
let plateBottom = NSColor(srgbRed: 0.129, green: 0.129, blue: 0.141, alpha: 1)
let branchColor = NSColor(srgbRed: 1, green: 1, blue: 1, alpha: 1)

// MARK: - Drawing

/// Unit-square, y-down mark coordinates onto an AppKit y-up canvas.
func canvasPoint(_ point: MarkPoint, size: CGFloat) -> NSPoint {
    NSPoint(x: CGFloat(point.x) * size, y: (1 - CGFloat(point.y)) * size)
}

func platePath(size: CGFloat) -> NSBezierPath {
    let path = NSBezierPath()
    for (index, point) in Mark.plateOutline().enumerated() {
        let canvas = canvasPoint(point, size: size)
        if index == 0 {
            path.move(to: canvas)
        } else {
            path.line(to: canvas)
        }
    }
    path.close()
    return path
}

/// One even-odd path holding all three rings, so each hollow core stays clear.
func ringsPath(size: CGFloat) -> NSBezierPath {
    let path = NSBezierPath()
    path.windingRule = .evenOdd
    for node in Mark.nodes {
        let center = canvasPoint(node, size: size)
        for radius in [Mark.outerRadius, Mark.innerRadius] {
            let scaled = CGFloat(radius) * size
            path.appendOval(in: NSRect(
                x: center.x - scaled,
                y: center.y - scaled,
                width: scaled * 2,
                height: scaled * 2
            ))
        }
    }
    return path
}

/// Lines from each ring's outer edge to the fork. Round caps tuck the ends back
/// under the ring stroke without reaching into the hollow core.
func branchLinesPath(size: CGFloat) -> NSBezierPath {
    let path = NSBezierPath()
    path.lineWidth = CGFloat(Mark.strokeWidth) * size
    path.lineCapStyle = .round
    path.lineJoinStyle = .round
    for node in Mark.nodes {
        let start = Mark.edgePoint(from: node, towards: Mark.fork)
        path.move(to: canvasPoint(start, size: size))
        path.line(to: canvasPoint(Mark.fork, size: size))
    }
    return path
}

func iconPNG(pixelSize px: Int) -> Data? {
    guard let rep = NSBitmapImageRep(
        bitmapDataPlanes: nil,
        pixelsWide: px,
        pixelsHigh: px,
        bitsPerSample: 8,
        samplesPerPixel: 4,
        hasAlpha: true,
        isPlanar: false,
        colorSpaceName: .deviceRGB,
        bytesPerRow: 0,
        bitsPerPixel: 0
    ) else { return nil }
    rep.size = NSSize(width: px, height: px)

    NSGraphicsContext.saveGraphicsState()
    defer { NSGraphicsContext.restoreGraphicsState() }
    guard let ctx = NSGraphicsContext(bitmapImageRep: rep) else { return nil }
    NSGraphicsContext.current = ctx
    ctx.shouldAntialias = true
    ctx.imageInterpolation = .high

    let size = CGFloat(px)

    // Ink plate, slightly lighter at the top.
    let plate = platePath(size: size)
    if let gradient = NSGradient(starting: plateBottom, ending: plateTop) {
        gradient.draw(in: plate, angle: 90)
    } else {
        plateBottom.setFill()
        plate.fill()
    }

    // White Y branch.
    branchColor.setStroke()
    branchLinesPath(size: size).stroke()
    branchColor.setFill()
    ringsPath(size: size).fill()

    let tagged = rep.converting(to: .sRGB, renderingIntent: .default) ?? rep
    return tagged.representation(using: .png, properties: [:])
}

// MARK: - Main

func warn(_ message: String) {
    if let data = (message + "\n").data(using: .utf8) {
        FileHandle.standardError.write(data)
    }
}

let fm = FileManager.default
if !fm.fileExists(atPath: assetPath) {
    try fm.createDirectory(atPath: assetPath, withIntermediateDirectories: true)
}

for (filename, px) in sizes {
    guard let data = iconPNG(pixelSize: px) else {
        warn("failed to encode \(filename)")
        continue
    }
    let url = URL(fileURLWithPath: "\(assetPath)/\(filename)")
    try data.write(to: url)
    let bytes = ByteCountFormatter.string(fromByteCount: Int64(data.count), countStyle: .file)
    print("  \(filename.padding(toLength: 20, withPad: " ", startingAt: 0))  \(px)×\(px)  \(bytes)")
}

print("")
print("Wrote \(sizes.count) icon PNGs to \(assetPath)")

File: scripts/generate-icon.py
#!/usr/bin/env python3
"""Generate GitRelay AppIcon PNGs (Linux-friendly mirror of generate-icon.swift)."""

from __future__ import annotations

import math
from pathlib import Path

from PIL import Image, ImageChops, ImageDraw

ASSET_PATH = Path("gitrelay/Assets.xcassets/AppIcon.appiconset")

SIZES = [
    ("icon_16.png", 16),
    ("icon_16@2x.png", 32),
    ("icon_32.png", 32),
    ("icon_32@2x.png", 64),
    ("icon_128.png", 128),
    ("icon_128@2x.png", 256),
    ("icon_256.png", 256),
    ("icon_256@2x.png", 512),
    ("icon_512.png", 512),
    ("icon_512@2x.png", 1024),
]

STROKE_WIDTH = 0.044
NODE_RADIUS = 0.072
OUTER_RADIUS = NODE_RADIUS + STROKE_WIDTH / 2
INNER_RADIUS = NODE_RADIUS - STROKE_WIDTH / 2
FORK = (0.5, 0.505)
NODES = [(0.5, 0.69), (0.293, 0.283), (0.707, 0.283)]
PLATE_EXPONENT = 5.0
PLATE_SAMPLE_COUNT = 720

PLATE_TOP = (52, 52, 55)
PLATE_BOTTOM = (33, 33, 36)
BRANCH = (255, 255, 255, 255)


def signed_power(value: float, exponent: float) -> float:
    magnitude = abs(value) ** exponent
    return -magnitude if value < 0 else magnitude


def plate_outline(sample_count: int = PLATE_SAMPLE_COUNT) -> list[tuple[float, float]]:
    exponent = 2 / PLATE_EXPONENT
    points: list[tuple[float, float]] = []
    for index in range(sample_count):
        angle = 2 * math.pi * index / sample_count
        points.append(
            (
                0.5 + 0.5 * signed_power(math.cos(angle), exponent),
                0.5 + 0.5 * signed_power(math.sin(angle), exponent),
            )
        )
    return points


def edge_point(node: tuple[float, float], target: tuple[float, float]) -> tuple[float, float]:
    dx = target[0] - node[0]
    dy = target[1] - node[1]
    length = math.hypot(dx, dy)
    if length <= 0:
        return node
    scale = OUTER_RADIUS / length
    return (node[0] + dx * scale, node[1] + dy * scale)


def canvas_point(point: tuple[float, float], size: int) -> tuple[float, float]:
    return point[0] * size, (1 - point[1]) * size


def ellipse_bbox(center: tuple[float, float], radius: float) -> list[float]:
    return [
        center[0] - radius,
        center[1] - radius,
        center[0] + radius,
        center[1] + radius,
    ]


def render_icon(pixels: int) -> Image.Image:
    size = pixels
    outline = [canvas_point(point, size) for point in plate_outline()]

    plate_mask = Image.new("L", (size, size), 0)
    ImageDraw.Draw(plate_mask).polygon(outline, fill=255)

    gradient = Image.new("RGBA", (size, size), (0, 0, 0, 0))
    grad_draw = ImageDraw.Draw(gradient)
    for y in range(size):
        t = 1 - y / max(size - 1, 1)
        color = tuple(
            int(PLATE_BOTTOM[i] + (PLATE_TOP[i] - PLATE_BOTTOM[i]) * t) for i in range(3)
        ) + (255,)
        grad_draw.line([(0, y), (size, y)], fill=color)

    image = Image.new("RGBA", (size, size), (0, 0, 0, 0))
    image.paste(gradient, mask=plate_mask)

    branch = Image.new("RGBA", (size, size), (0, 0, 0, 0))
    draw = ImageDraw.Draw(branch)
    stroke = max(1, round(STROKE_WIDTH * size))

    for node in NODES:
        start = canvas_point(edge_point(node, FORK), size)
        end = canvas_point(FORK, size)
        draw.line([start, end], fill=BRANCH, width=stroke)

    outer_mask = Image.new("L", (size, size), 0)
    inner_mask = Image.new("L", (size, size), 0)
    outer_draw = ImageDraw.Draw(outer_mask)
    inner_draw = ImageDraw.Draw(inner_mask)
    for node in NODES:
        center = canvas_point(node, size)
        outer_draw.ellipse(ellipse_bbox(center, OUTER_RADIUS * size), fill=255)
        inner_draw.ellipse(ellipse_bbox(center, INNER_RADIUS * size), fill=255)
    ring_mask = ImageChops.subtract(outer_mask, inner_mask)
    rings = Image.new("RGBA", (size, size), BRANCH)
    branch.paste(rings, mask=ring_mask)

    return Image.alpha_composite(image, branch)


def main() -> None:
    ASSET_PATH.mkdir(parents=True, exist_ok=True)
    for filename, px in SIZES:
        out = ASSET_PATH / filename
        render_icon(px).save(out, format="PNG")
        print(f"  {filename:<20} {px}x{px}  {out.stat().st_size:,} bytes")
    print(f"\nWrote {len(SIZES)} icon PNGs to {ASSET_PATH}")


if __name__ == "__main__":
    main()
